### ISSUE TITLE
Watchpath will not trigger when the glob is early in the pattern

### ISSUE THREAD

**Issue by robsyme** [nextflow-io/nextflow#2721]
### Expected behavior and actual behavior

I would expect that we can introduce the glob "*" in a watchPath pattern at a location other than the end of the path.

For example, `Channel.watchPath("watch/*/outputs/*.txt")` where the first glob does not appear at the final directory (the `outputs` directory appears afterwards) will not detect matching events.

### Steps to reproduce the problem

In main.nf:

```groovy
// This will work perfectly
Channel.watchPath("watch/projectA/*/*.txt").view { "Glob at end of pattern: $it" }

// This will not detect any events
Channel.watchPath("watch/*/outputs/*.txt").view { "Glob before end of pattern: $it" }
```

Set up the directories, run the workflow:

```bash
mkdir -p watch/project{A,B,C}/outputs
nextflow run .
```
... and in a separate shell create three files that match the second pattern.

```bash
# ABC permuted to CAB to show it is not an ordering problem
touch watch/project{C,A,B}/outputs/test.txt
```

But only the `projectA/outputs/test.txt` (from the first `watchPath`) is detected:

```
Launching `./main.nf` [gloomy_cray] - revision: 76651a7ada
Glob at end of pattern: <basedir>/watch/projectA/outputs/test.txt
```

The most relevant lines from the `.nextflow.log` are the `DirWatcher` logs:

```
Mar.-14 11:11:30.436 [main] DEBUG nextflow.file.DirWatcher - Watch service for path=watch/projectA/; syntax=glob; pattern=*/*.txt; skipHidden=true; events=create
Mar.-14 11:11:30.518 [main] DEBUG nextflow.file.DirWatcher - Watch service for path=watch/; syntax=glob; pattern=*/outputs/*.txt; skipHidden=true; events=create
```

### Environment 

* Nextflow version: 21.10.6
* Java version: openjdk 11.0.2 2019-01-15
* Operating system: macOS, Linux
* Bash version: (use the command `zsh 5.8 (x86_64-apple-darwin21.0)` and `GNU bash, version 4.2.46`)

### Additional context

I raised this in the Nextflow Slack channel. Paolo offered that it might be a shared filesystem problem, but I've excluded that by running in other filesystems and looking closely at the `DirWatcher` code.

### Specifics

In this section I'm only considering `Channel.watchPath("watch/*/outputs/*.txt")` - the failing case.

The problem stems from the fact that the `DirWatcher` class can only match a single `PathMatcher`. It matches via the  `PathMatcher dirMatcher` field.

When the `DirWatcher` instance is created, the log (above) shows that a new `PathMatcher` is created, watching the `watch/` directory with pattern "`*/outputs/*.txt`".

This PathMatcher is registered with the filesystem for watching, but it is registered on the `watch` directory. When new items are created in the `watch` directory, the DirWatcher will check if that new items matches the pattern (`*/outputs/*.txt`) but of course no file or folder created in that `watch` directory could match that pattern - only files created in the `outputs` directory could match that pattern, but we are only watching the parent (`watch`) directory.

My suggestion is to change the `dirMatcher` field to a List: `List<PathMatcher> dirMatchers` that get populated in the constructor:

```groovy
pattern
.findIndexValues { it == '/'}
.each { idx ->
    def prefix = pattern.substring(0, idx as int)
    dirRule = "$syntax:${prefix}"
    dirMatchers << FileHelper.getPathMatcherFor(dirRule, base.fileSystem)
}
```

This way we are creating two pathMatchers: one `watch`, and one `watch/*`

In the `apply0` method, we would have to register each of these pathMatchers:

```groovy
for (dm in dirMatchers) {
    register(base,dm)
}
```

And then when iterating over new events, check if `kind == ENTRY_CREATE` and `Files.isDirectory(target)` and if so, check if the new directory matches on of our pathMatchers, and if it does, register it:

```groovy
if( kind == ENTRY_CREATE && dirMatchers && Files.isDirectory(target, LinkOption.NOFOLLOW_LINKS) ) {
    def targetdir = base.relativize(target)
    def dm = dirMatchers.find { dm -> dm.matches(targetdir) }
    if( dm ) {
        register(target, dm)
    }
}
```

I can send a PR with tests and a fix if that would be helpful.

**Comment by robsyme** [comment]
I've made the changes in the `[dirwatch-deeper](https://github.com/robsyme/nextflow/tree/dirwatch-deeper)` branch in my repo. 

The only problem remaining is that if a watched folder is deleted, the Channel closes.  For example:

Given a `Channel.watchPath("watch/*/outputs/*.txt")`, and a tree:

```
watch
├── bar
│   └── outputs
└── foo
    └── outputs
```

If the channel is watching and the `bar` directory is removed, the Channel closes, even though the `foo` path may still be available.

My feeling is that the `watchPath` channels should stay open indefinitely, but I might be in the minority.

**Comment by pditommaso** [comment]
Any chance to make a PR even better including a [unit test](https://github.com/nextflow-io/nextflow/blob/4e9189ac8e5fcead19b509f0905dc00607783fe1/modules/nextflow/src/test/groovy/nextflow/file/DirWatcherTest.groovy#L33-L33) highlighting the problem?

**Comment by robsyme** [comment]
I thought you might ask, so I already [included one](https://github.com/robsyme/nextflow/commit/1c5fcc3676ea34e7546695e67981d822d9179cb6#diff-e09acc43f9b2df87cc8e849c4403fb47a7ef18bcd013d4cce6bc05e41ea8024c) ;)

**Comment by pditommaso** [comment]
ahah, so the test pass or fail? in any case make a pull request, so we'll see it in action

**Comment by robsyme** [comment]
The unit test fails on `master`

```
Condition not satisfied:

results.size() == 3
|       |      |
[]      0      false
```

... and passes with updated DirWatcher class

```
should watch when glob is early in pattern | 10.527s | passed
```

I'll submit the PR.

In the meantime, do you have strong opinions about letting users keep the channel open, even if one of the watched directories is deleted?

**Comment by robsyme** [comment]
Closing, as this issue is resolved with DirWatcherV2, available in `22.06.0-edge` or newer.
